HTTP_API.md

```diff
@@ -0,0 +1,952 @@
+    Or there was an unexpected error. The body is an error message.
+
+    ###### example
+    ```
+    $ curl localhost:8181/api/v1/grafana_datasource_url
+    {"grafanaDatasourceUrl":"http://localhost:8080"}
+    ```
+
+
+* #### `HealthGetHandler`
+
+    ###### synopsis
+    Returns whether or not the Grafana datasource and Grafana dashbaord
+    that Container JFR is configured with are running properly.
+
+    ###### request
+    `GET /health`
+
+    ###### response
+    `200` - The body is
+    `{"datasourceAvailable":$DATASOURCE_AVAILABLE,"dashboardAvailable":$DASHBOARD_AVAILABLE}`.
+
+    `$DATASOURCE_AVAILABLE` is `true` if  Container JFR is configured with a
+    Grafana datasource and that datasource responds to a `GET` request
+    with a `200`, and it is `false` otherwise.
+
+    `$DASHBOARD_AVAILABLE` is `true` if  Container JFR is configured with a
+    Grafana dashboard and that dashboard responds to a `GET` request
+    with a `200`, and it is `false` otherwise.
+
+    `500` - There was an unexpected error. The body is an error message.
+
+    ###### example
+    ```
+    $ curl localhost:8181/health
+    {"dashboardAvailable":false,"datasourceAvailable":false}
+    ```
+
+
+* #### `StaticAssetsGetHandler`
+
+    ###### synopsis
+    Returns a static asset of the Container JFR web client.
+
+    ###### request
+    `GET /*`
+
+    The path should be the directory path of the desired asset,
+    assuming a root of
+    `target/assets/app/resources/com/redhat/rhjmc/containerjfr/net/web/`.
+
+    ###### response
+
+    `200` - The body is the requested asset. Note that if the requested asset
+    does not exist or the path is invalid, the request will not be handled here
+    and instead  will be routed to
+    [`WebClientAssetsGetHandler`](#WebClientAssetsGetHandler).
+
+    `500` - There was an unexpected error. The body is an error message.
+
+    ###### example
+    ```
+    $ curl localhost:8181/images/pfbg_992@2x.jpg --output pfbg_992@2x.jpg
+      % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
+                                     Dload  Upload   Total   Spent    Left  Speed
+    100  574k  100  574k    0     0  43.1M      0 --:--:-- --:--:-- --:--:-- 43.1M
+    ```
+
+
+* #### `TargetsGetHandler`
+
+    ###### synopsis
+    Scans for and returns discoverable target JVMs.
+
+    ###### request
+    `GET /api/v1/targets`
+
+    ###### response
+    `200` - The body consists of a comma-seperated list of targets,
+    enclosed in square brackets.
+
+    The format for a target is
+    `{"connectUrl":"$CONNECT_URL","alias":"$ALIAS"}`.
+
+    So a list with no recordings looks like `[]`,
+    a list with one recording looks like `[{...}]`,
+    and a list with two recordings looks like `[{...},{...}]`.
+
+    `401` - User authentication failed. The body is an error message.
+
+    `407` - JMX authentication failed. The body is an error message.
```

Comment:
I'm adding this note here but it applies to all `407`s, or maybe this needs a separate explanation. When a `407` is sent, the response will also include an `X-JMX-Authenticate` header, currently always with the value `Basic` but in the future this could also be `Bearer` or other authentication schemes. This indicates to the client the form of authentication that is expected/accepted.

Reply:
Do the `401` need a similar note, or is it only because this is a custom header?

Reply:
Yea, this is a custom header so it needs additional documentation. There is a similar dynamic with the `401 / Authenticate / Authorization` which I guess we should also document, since we use a custom `X-WWW-Authenticate` instead of the standard `WWW-Authenticate`.